crates/gooey-widgets/src/input.rs

```diff
@@ -0,0 +1,243 @@
+use gooey_core::style::Color;
+use gooey_core::{AnyCallback, Callback, Value, Widget};
+
+use crate::State;
+
+#[derive(Debug, Default, Clone, Widget)]
+#[widget(authority = gooey)]
+#[must_use]
+pub struct Input {
+    pub value: Value<String>,
+    pub on_update: Option<Callback<String>>,
+}
+
+impl Input {
+    pub fn new(value: impl Into<Value<String>>) -> Self {
+        Self {
+            value: value.into(),
+            ..Self::default()
+        }
+    }
+
+    pub fn value(mut self, value: impl Into<Value<String>>) -> Self {
+        self.value = value.into();
+        self
+    }
+
+    pub fn on_update<CB: AnyCallback<String>>(mut self, cb: CB) -> Self {
+        self.on_update = Some(Callback::new(cb));
+        self
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct InputTransmogrifier;
+
+#[cfg(feature = "web")]
+mod web {
+    use futures_util::StreamExt;
+    use gooey_core::style::DynamicStyle;
+    use gooey_core::{Value, WidgetTransmogrifier};
+    use gooey_web::WebApp;
+    use wasm_bindgen::prelude::Closure;
+    use wasm_bindgen::JsCast;
+    use web_sys::HtmlInputElement;
+
+    use crate::input::{Input, InputTransmogrifier};
+
+    impl WidgetTransmogrifier<WebApp> for InputTransmogrifier {
+        type Widget = Input;
+
+        fn transmogrify(
+            &self,
+            widget: &Self::Widget,
+            _style: DynamicStyle,
+            _context: &<WebApp as gooey_core::Frontend>::Context,
+        ) -> <WebApp as gooey_core::Frontend>::Instance {
+            // TODO apply style
+            let label = widget.value.clone();
+            let on_update = widget.on_update.clone();
+            let document = web_sys::window()
+                .expect("no window")
+                .document()
+                .expect("no document");
+            let input = document
+                .create_element("input")
+                .expect("failed to create input")
+                .dyn_into::<HtmlInputElement>()
+                .expect("incorrect element type");
+
+            label.map_ref(|label| input.set_inner_text(label));
+
+            if let Value::Dynamic(label) = label {
+                let input = input.clone();
+                wasm_bindgen_futures::spawn_local(async move {
+                    let mut label = label.into_stream();
+                    while let Some(new_label) = label.next().await {
+                        input.set_inner_text(&new_label);
+                    }
+                });
+            }
+
+            if let Some(mut on_update) = on_update {
+                let closure = Closure::new(move || {
+                    on_update.invoke(String::new());
```

Comment:
Unsure how to access the `input` element here to get the value, or do I need to go through `event.target.value`

Reply:
WASM closures don't need to be sync. You can just move the element into it, in theory.

Reply:
yeah, but I need to attach this closure to the `input`, so I cannot move it.

```rs
input.on_update(move || input /*...*/) // input referenced after move
```

Reply:
I somehow missed this comment. I think you can just clone any Node to get a new reference to the underlying object. But, I can't find any definitive source of that knowledge or any existing code i wrote that uses it.

Reply:
I'll try that when I'm back from htmx :D

Reply:
No rush!